File: src/scripts/protocols/lib/data-reader.ts
import 'core-js/features/reflect/construct'
import jsbi from 'jsbi'

const { BigInt } = jsbi
const big32 = BigInt(32)

const readUInt64LE = (buf: Buffer, byteOffset: number) => {
  const wh = BigInt(buf.readUInt32LE(byteOffset + 4))
  const wl = BigInt(buf.readUInt32LE(byteOffset))
  return jsbi.add(jsbi.leftShift(wh, big32), wl)
}

export class DataReader {
  private _buf: Buffer
  public pos = 0

  constructor(buf: Buffer) {
    this._buf = buf
  }

  skip(length: number) {
    this.pos += length
    return this
  }

  u8() {
    const ret = this._buf.readUInt8(this.pos)
    this.pos += 1
    return ret
  }

  u16le() {
    const ret = this._buf.readUInt16LE(this.pos)
    this.pos += 2
    return ret
  }

  u32le() {
    const ret = this._buf.readUInt32LE(this.pos)
    this.pos += 4
    return ret
  }

  u64le() {
    const ret = readUInt64LE(this._buf, this.pos)
    this.pos += 8
    return ret
  }

  i8() {
    const ret = this._buf.readInt8(this.pos)
    this.pos += 1
    return ret
  }

  i16le() {
    const ret = this._buf.readInt16LE(this.pos)
    this.pos += 2
    return ret
  }

  i32le() {
    const ret = this._buf.readInt32LE(this.pos)
    this.pos += 4
    return ret
  }

  f32le() {
    const ret = this._buf.readFloatLE(this.pos)
    this.pos += 4
    return ret
  }

  f64le() {
    const ret = this._buf.readDoubleLE(this.pos)
    this.pos += 8
    return ret
  }

  zstring() {
    const idx = Array.prototype.indexOf.call(this._buf, 0, this.pos)

    if (idx == -1) {
      return this.lstring(this._buf.length - this.pos)
    }

    return this.lstring(idx - this.pos + 1)
  }

  lstring(length: number) {
    const ret = this._buf.toString('binary', this.pos, this.pos + length)
    this.pos += length
    return ret
  }

  data() {
    const ret = this._buf.slice(this.pos)
    this.pos += ret.byteLength
    return ret
  }
}

File: data/protocols/lib/DataReader.lua
---@meta

local Object = require("lib/classic")

---@class DataReader
local DataReader = Object:extend()

---@param buf string
function DataReader:new(buf)
  self.buf = buf
  self.pos = 1
end

function DataReader:is_end()
  return self.pos > #self.buf
end

---@param length integer
function DataReader:skip(length)
  self.pos = self.pos + length
end

---@return integer
function DataReader:u8()
  local val, pos = string.unpack("I1", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u16le()
  local val, pos = string.unpack("<I2", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u16be()
  local val, pos = string.unpack(">I2", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u32le()
  local val, pos = string.unpack("<I4", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u32be()
  local val, pos = string.unpack(">I4", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u64le()
  local val, pos = string.unpack("<I8", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:u64be()
  local val, pos = string.unpack(">I8", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i8()
  local val, pos = string.unpack("i1", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i16le()
  local val, pos = string.unpack("<i2", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i16be()
  local val, pos = string.unpack(">i2", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i32le()
  local val, pos = string.unpack("<i4", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i32be()
  local val, pos = string.unpack(">i4", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i64le()
  local val, pos = string.unpack("<i8", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return integer
function DataReader:i64be()
  local val, pos = string.unpack(">i8", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return number
function DataReader:f32le()
  local val, pos = string.unpack("<f", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return number
function DataReader:f32be()
  local val, pos = string.unpack(">f", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return number
function DataReader:f64le()
  local val, pos = string.unpack("<d", self.buf, self.pos)
  self.pos = pos
  return val
end

---@return number
function DataReader:f64be()
  local val, pos = string.unpack(">d", self.buf, self.pos)
  self.pos = pos
  return val
end

---@param length integer
---@return string
function DataReader:string(length)
  local val, pos = string.unpack("c" .. length, self.buf, self.pos)
  self.pos = pos
  return val
end

---@return string
function DataReader:lstring()
  local val, pos = string.unpack("s1", self.buf, self.pos)
  self.pos = pos
  return val
end

---@param delimiter? string
---@return string
function DataReader:zstring(delimiter)
  delimiter = delimiter or "\0"
  local pos = self.buf:find(delimiter, self.pos, true)

  if pos == nil then
    pos = #self.buf + 1
  end

  local val = self.buf:sub(self.pos, pos - #delimiter)
  self.pos = pos + #delimiter
  return val
end

---@return string
function DataReader:data()
  local data = self.buf:sub(self.pos)
  self.pos = #self.buf
  return data
end

return DataReader
